pkg/repository/oci_watcher.go

```diff
@@ -0,0 +1,225 @@
+	"github.com/kubebb/core/pkg/helm"
+	"github.com/kubebb/core/pkg/utils"
+)
+
+var _ IWatcher = (*OciWatcher)(nil)
+
+func NewOciWatcher(
+	instance *v1alpha1.Repository,
+	c client.Client,
+	ctx context.Context,
+	logger logr.Logger,
+	duration time.Duration,
+	cancel context.CancelFunc,
+	scheme *runtime.Scheme,
+	fm map[string]v1alpha1.FilterCond,
+) IWatcher {
+	result := &OciWatcher{
+		instance:  instance,
+		logger:    logger,
+		duration:  duration,
+		cancel:    cancel,
+		scheme:    scheme,
+		repoName:  instance.NamespacedName(),
+		filterMap: fm,
+	}
+
+	// Common Action in the watcher needs client and context to function
+	result.c = c
+	result.ctx = ctx
+	return result
+}
+
+type OciWatcher struct {
+	CommonAction
+	cancel   context.CancelFunc
+	instance *v1alpha1.Repository
+	duration time.Duration
+	repoName string
+
+	username  string
+	password  string
+	logger    logr.Logger
+	scheme    *runtime.Scheme
+	filterMap map[string]v1alpha1.FilterCond
+}
+
+func (c *OciWatcher) Start() error {
+	username, password, err := Start(c.ctx, c.instance, c.duration, c.repoName, c.c, c.logger)
+	if err != nil {
+		return err
+	}
+	c.username = username
+	c.password = password
+
+	if err := helm.RepoAdd(c.ctx, c.logger, c.repoName, c.instance.Spec.URL, c.username, c.password, c.duration/2); err != nil {
+		c.logger.Error(err, "Failed to add repository")
+		return err
+	}
+	if err := helm.RepoUpdate(c.ctx, c.logger, c.repoName, c.duration/2); err != nil {
+		c.logger.Error(err, "Failed to update repository")
+		return err
+	}
+
+	go wait.Until(c.Poll, c.duration, c.ctx.Done())
+	return nil
+}
+
+func (c *OciWatcher) Stop() {
+	c.logger.Info("Delete Or Update Repository, stop watcher")
+	if err := helm.RepoRemove(c.ctx, c.logger, c.repoName); err != nil {
+		c.logger.Error(err, "Failed to remove repository")
+	}
+	c.cancel()
+}
+
+// Poll the components
+func (c *OciWatcher) Poll() {
+	c.logger.Info("OCI poll")
+	now := metav1.Now()
+	readyCond := getReadyCond(now)
+	syncCond := getSyncCond(now)
+
+	if err := helm.RepoUpdate(c.ctx, c.logger, c.repoName, c.duration/2); err != nil {
+		c.logger.Error(err, "Failed to update repository")
+		return
+	}
+	entryName := utils.GetOCIEntryName(c.instance.Spec.URL)
+	cfg, err := ctrl.GetConfig()
+	if err != nil {
+		c.logger.Error(err, "Cannot get config")
+		return
+	}
+	ns := c.instance.GetNamespace()
+	getter := genericclioptions.ConfigFlags{
+		APIServer:   &cfg.Host,
+		CAFile:      &cfg.CAFile,
+		BearerToken: &cfg.BearerToken,
+		Namespace:   &ns,
+	}
+
+	h, err := helm.NewHelmWrapper(&getter, ns, c.logger)
+	if err != nil {
+		c.logger.Error(err, "Cannot create a new helm wrapper")
+	}
+	info, err := h.Pull(c.logger, c.instance.Spec.URL)
+	if err != nil {
+		c.logger.Error(err, "Cannot pull chart")
+		return
+	}
+	versions := info.Metadata
+
+	// filter component and its versions
+	// _, keep := v1alpha1.Match(c.filterMap, v1alpha1.Filter{Name: entryName, Versions: versions.Version})
```

Comment:
why leave these comments here?

Reply:
These comments were in the HTTPWatcher and it seems like it is used to filter the components. I put a copy in the OCIWatcher in case we will need it in the future, but it is also OK to delete them!